daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,217 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import java.io.File
+
+import org.apache.daffodil.compiler.Compiler
+import org.apache.daffodil.dsom.{Choice, ComplexTypeBase, ElementBase, GroupRef, ModelGroup, Root, Sequence, SimpleTypeBase, Term}
+
+/**
+ * A class designed to walk the internal representation of a DFDL Schema File.
+ *
+ * There are numerous internal event handlers to be defined by the implementing class; each
+ * corresponds to "encountering" a particular element type in the DFDL Schema file.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ * @param schemaFile an input DFDL Schema file to parse
+ * @tparam T the return type of walkDSOMSchema and onWalkEnd.  This should be a class used to
+ *           represent the structure of the Schema.  You could always make it Unit if you don't
+ *           want these methods to return anything.
+ */
+abstract class AbstractDSOMWalker[T](schemaFile: File) {
+  /**
+   * The root element of the DFDL Schema.  This will be the starting point for the traversal
+   */
+  private final val schemaSetRoot: Root = Compiler().compileFile(schemaFile).sset.root
+
+  /**
+   * Auxiliary constructor to initialize this class with a path to a schema File rather than
+   * a File object directly
+   * @param pathToSchemaFile path to the DFDL Schema file desired to be parsed
+   */
+  def this(pathToSchemaFile: String) = this(new File(pathToSchemaFile))
+
+  /**
+   * Method to be called on the beginning of the traversal.  It is recommended to add some
+   * sort of wrapper element to a stack if you're doing a typical stack-based traversal.
+   * @param root the root element of the DFDL Schema
+   */
+  protected def onWalkBegin(root: Root): Unit
+
+  /**
+   * Method to be called when the traversal concludes.  It is recommended to put any post-processing
+   * and anything to tidy up the stack or the result here.
+   * @param root the root element of the DFDL Schema
+   * @return the result of the traversal
+   */
+  protected def onWalkEnd(root: Root): T
+
+  /**
+   * Method to be called when a *sequence* element is encountered.
+   * This is not really your array type; the typical "array" you'll see in an infoset
+   * is a special kind of Element, so it will be encountered
+   * in onElementBegin
+   * @param sequenceElement the sequence element
+   */
+  protected def onSequenceBegin(sequenceElement: Sequence): Unit
```

Comment:
I'm a little hesitant to give a walker implementation direct access to all of our internal schema compontents. It makes it much more difficult to make changes to these SchemaComponents since we need to be concerned about backwards compatibilty. We sort of have that problem with our InfosetOutputters, I'd like to not repeat that mistake.

I'm wondering if it would be possible to come up with a reasonable list of properties that DSOM walkers actually need and only make those avaialble? I suspect this would be a pretty small list llike namespace, name, type, min/maxOccurs, and maybe a couple more properties?

Perhaps your SchemaNode is the correct abstraction? Maybe fore each walker event, the we should be creating a SchemaNode and passing that in? Then we only have to worry about backwards compatibility with SchemaNodes, which should be significantly easier than our SchemaComponents?

Reply:
The abstract walker can define a set of view mixin traits that we mix into each walked object. The traits may be specific to each walked DSOM object as it has to expose different things from them perhaps.

The walker API would be defined in terms of these traits. The traits would have public members that are the things we choose to formally expose (and support) on the walker API.

Reply:
I'd suggest the names should end with "View" so ComplexTypeView would be the trait mixed into ComplexTypeBase giving the walker access to its public members, which would just be pass through. 

To cheat users could always downcast to ComplexTypeBase, but I think that's glaring enough that we can live with it.

Reply:
There needs to be a way to "register" a walker. I.e., a Daffodil API call where the walker is passed to Daffodil, so that it can be invoked. Or if the control pattern is that the walker takes control, it's really just a hook where daffodil informs the walker that the DSOM tree is ready to be walked. This should allow for many external walkers to be registered, not just one. They should be called in turn. 

Then all the NiFi-specific logic, tests, etc. all goes into distinct separate code that is not part of Daffodil. It becomes an application of Daffodil's libraries within NiFi code.

Reply:
This all makes a lot of sense, hopefully I took care of it well in the latest version.

I wound up defining a series of traits ending in View as Mike suggested, such as TermView and SequenceView.  Everything in RecordWalker and the helper SchemaNode (which I renamed to RecordSchemaNode to avoid confusion, it should _only_ be used for the NIFI RecordSchema, as it has an attribute that is evaluated to a NIFI Record Simple type, now uses all of these traits.

The only "internal" item the traits can access that isn't a primitive or a String is the Enum PrimType class; I decided to keep it since it was highly useful when it came to simple types.  I hope leaving this in won't cause any issues.

Within these traits I also defined a new "walkDSOM" method; hence moving all of those walkerHelper methods out of the abstract class and into the new traits.